msi/createmsi.py

```diff
@@ -78,78 +35,177 @@ def __init__(self, dirs, files):
+            'Name': 'SourceDir',
+            })
+        progfiledir = ET.SubElement(targetdir, 'Directory', {
+            'Id' : self.progfile_dir,
+            })
+        installdir = ET.SubElement(progfiledir, 'Directory', {
+            'Id': 'INSTALLDIR',
+            'Name': 'Meson'})
+
+        ET.SubElement(product, 'Property', {
+            'Id': 'WIXUI_INSTALLDIR',
+            'Value': 'INSTALLDIR',
+            })
+        ET.SubElement(product, 'UIRef', {
+            'Id': 'WixUI_FeatureTree',
+            })
+        for sd in self.staging_dirs:
+            assert(os.path.isdir(sd))
+        top_feature = ET.SubElement(product, 'Feature', {
+            'Id': 'Complete',
+            'Title': 'Meson ' + self.version,
+            'Description': 'The complete package',
+            'Display': 'expand',
+            'Level': '1',
+            'ConfigurableDirectory': 'INSTALLDIR',
+            })
+        for sd in self.staging_dirs:
+            nodes = {}
+            for root, dirs, files in os.walk(sd):
                 cur_node = Node(dirs, files)
                 nodes[root] = cur_node
-            ofile.write(xml_templ % (self.guid, self.update_guid, self.version, self.version,
-                                     self.platform_str, self.progfile_dir))
-            self.component_num = 0
-            self.create_xml(nodes, ofile, self.staging_dir)
-            for i in range(self.component_num):
-                comp_ref_xml += comp_ref_templ % ('ApplicationFiles%d' % i)
-            ofile.write(xml_footer_templ % comp_ref_xml)
-
-    def create_xml(self, nodes, ofile, root):
-        cur_node = nodes[root]
+            self.create_xml(nodes, sd, installdir, sd)
+            self.build_features(nodes, top_feature, sd)
+        ET.ElementTree(self.root).write(self.main_xml, encoding='utf-8',xml_declaration=True)
+        # ElementTree can not do prettyprinting so do it manually
+        import xml.dom.minidom
+        doc = xml.dom.minidom.parse(self.main_xml)
+        with open(self.main_xml, 'w') as of:
+            of.write(doc.toprettyxml())
+
+    def build_features(self, nodes, top_feature, staging_dir):
+        feature = ET.SubElement(top_feature, 'Feature',  self.feature_properties[staging_dir])
+        for component_id in self.feature_components[staging_dir]:
+            ET.SubElement(feature, 'ComponentRef', {
+                'Id': component_id,
+                })
+        
+
+    def create_xml(self, nodes, current_dir, parent_xml_node, staging_dir):
+        cur_node = nodes[current_dir]
         if cur_node.files:
-            ofile.write("<Component Id='ApplicationFiles%d' Guid='%s' %s>\n" % (self.component_num, gen_guid(), self.component_platform))
+            component_id = 'ApplicationFiles%d' % self.component_num
+            comp_xml_node = ET.SubElement(parent_xml_node, 'Component', {
+                'Id': component_id,
+                'Guid': gen_guid(),
+                })
```

Comment:
__[flake8]__
* [E123] closing bracket does not match indentation of opening bracket's line

<!-- Comment by [SideCI](https://sideci.com) -->